### PR TITLE
Python: Add pickle support

### LINKED ISSUE #7644: Python: Pickle support
### Feature Request / Improvement

Right now, various classes in PyIceberg cannot be pickled. This is due `__new__` lacking a corresponding `__getnewargs__` in various places ([pickle docs](https://docs.python.org/3/library/pickle.html#object.__getnewargs__)). Pickle support is required by, e.g., Dask to serialise tasks as they are distributed to workers.

Here is a simple example:
```python
import pickle
from pyiceberg.expressions import And, EqualTo

ex = And(EqualTo('name', 'test'), EqualTo('value', 0))
pickle.loads(pickle.dumps(ex))
```

This results in the following exception:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: And.__new__() missing 2 required positional arguments: 'left' and 'right'
```

I'm happy to provide a fix, but I'd like to hear your thoughts first.

### Query engine

Other

### PR DESCRIPTION
**Pull request by grobgl**
Closes #7644

### CODE REVIEW

File: python/pyiceberg/types.py [hunk 1]
```diff
@@ -243,6 +249,9 @@ def __str__(self) -> str:
         req = "required" if self.required else "optional"
         return f"{self.field_id}: {self.name}: {req} {self.field_type}{doc}"
 
+    def __getnewargs__(self) -> Tuple[Optional[int], Optional[str], Optional[IcebergType], bool, Optional[str]]:
```
Comment (Fokko) [inline python/pyiceberg/types.py:252]:
```suggestion
    def __getnewargs__(self) -> Tuple[int, str, IcebergType, bool, Optional[str]]:
```

**Review by Fokko — APPROVED** [review]
Some small suggestions, but looks good! Thanks for working on the Dask support!

**Comment by Fokko** [comment]
Thanks for working on this @grobgl. Let me know if I can help in any way with the Dask support!

**Comment by grobgl** [comment]
Thanks for sorting this out! I was on leave for a while, hoping to get back on it soon.

**Pull request merged**